exchanges/subscription/subscription.go

```diff
@@ -1,92 +1,152 @@
+	ErrDuplicate      = errors.New("duplicate subscription")
 )
 
+// State tracks the status of a subscription channel
+type State uint8
+
 // Subscription container for streaming subscriptions
 type Subscription struct {
-	Enabled       bool                   `json:"enabled"`
-	Key           any                    `json:"-"`
-	Channel       string                 `json:"channel,omitempty"`
-	Pair          currency.Pair          `json:"pair,omitempty"`
-	Asset         asset.Item             `json:"asset,omitempty"`
-	Params        map[string]interface{} `json:"params,omitempty"`
-	State         State                  `json:"-"`
-	Interval      kline.Interval         `json:"interval,omitempty"`
-	Levels        int                    `json:"levels,omitempty"`
-	Authenticated bool                   `json:"authenticated,omitempty"`
+	Enabled       bool           `json:"enabled"`
+	Key           any            `json:"-"`
+	Channel       string         `json:"channel,omitempty"`
+	Pairs         currency.Pairs `json:"pairs,omitempty"`
+	Asset         asset.Item     `json:"asset,omitempty"`
+	Params        map[string]any `json:"params,omitempty"`
+	Interval      kline.Interval `json:"interval,omitempty"`
+	Levels        int            `json:"levels,omitempty"`
+	Authenticated bool           `json:"authenticated,omitempty"`
+	state         State
+	m             sync.RWMutex
 }
 
-// MarshalJSON generates a JSON representation of a Subscription, specifically for config writing
-// The only reason it exists is to avoid having to make Pair a pointer, since that would be generally painful
-// If Pair becomes a pointer, this method is redundant and should be removed
-func (s *Subscription) MarshalJSON() ([]byte, error) {
-	// None of the usual type embedding tricks seem to work for not emitting an nil Pair
-	// The embedded type's Pair always fills the empty value
-	type MaybePair struct {
-		Enabled       bool                   `json:"enabled"`
-		Channel       string                 `json:"channel,omitempty"`
-		Asset         asset.Item             `json:"asset,omitempty"`
-		Params        map[string]interface{} `json:"params,omitempty"`
-		Interval      kline.Interval         `json:"interval,omitempty"`
-		Levels        int                    `json:"levels,omitempty"`
-		Authenticated bool                   `json:"authenticated,omitempty"`
-		Pair          *currency.Pair         `json:"pair,omitempty"`
+// String implements Stringer, and aims to informatively and uniquely identify a subscription for errors and information
+// returns a string of the subscription key by delegating to MatchableKey.String() when possible
+// If the key is not a MatchableKey then both the key and an ExactKey.String() will be returned; e.g. 1137: spot MyTrades
+func (s *Subscription) String() string {
+	key := s.EnsureKeyed()
+	if k, ok := key.(MatchableKey); ok {
+		return k.String()
 	}
+	return fmt.Sprintf("%v: %s", key, ExactKey{s}.String())
```

Comment:
Looks like a potential race zone

Reply:
Yeah. Not particularly easy to get around, either.

Fixed 913fb2d